controllers/envoy/envoyconfigrevision_controller.go

```diff
@@ -308,14 +179,71 @@ func (r *EnvoyConfigRevisionReconciler) updateStatus(ctx context.Context, ecr *e
 	return nil
 }
 
-func calculateSecretsHash(resources map[string]envoy.Resource) string {
-	resourcesHasher := fnv.New32a()
-	common.DeepHashObject(resourcesHasher, resources)
-	return rand.SafeEncodeString(fmt.Sprint(resourcesHasher.Sum32()))
+func (r *EnvoyConfigRevisionReconciler) addFinalizer(ctx context.Context, ecr *envoyv1alpha1.EnvoyConfigRevision) error {
+	controllerutil.AddFinalizer(ecr, envoyv1alpha1.EnvoyConfigRevisionFinalizer)
+
+	// Update CR
+	if err := r.Client.Update(ctx, ecr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *EnvoyConfigRevisionReconciler) finalizeEnvoyConfigRevision(nodeID string) {
+	r.XdsCache.ClearSnapshot(nodeID)
+}
+
+func (r *EnvoyConfigRevisionReconciler) taintSelf(ctx context.Context, ecr *envoyv1alpha1.EnvoyConfigRevision, reason, msg string) error {
+	if !ecr.Status.Conditions.IsTrueFor(envoyv1alpha1.RevisionTaintedCondition) {
+		patch := client.MergeFrom(ecr.DeepCopy())
+		ecr.Status.Conditions.SetCondition(status.Condition{
+			Type:    envoyv1alpha1.RevisionTaintedCondition,
+			Status:  corev1.ConditionTrue,
+			Reason:  status.ConditionReason(reason),
+			Message: msg,
+		})
+		ecr.Status.Tainted = true
+
+		if err := r.Client.Status().Patch(ctx, ecr, patch); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func filterByAPIVersion(obj runtime.Object, version envoy.APIVersion) bool {
+	switch o := obj.(type) {
+	case *envoyv1alpha1.EnvoyConfigRevision:
+		if o.GetEnvoyAPIVersion() == version {
+			return true
+		}
+		return false
+
+	default:
+		return false
+	}
+}
+
+func filterByAPIVersionPredicate(version envoy.APIVersion,
+	filter func(runtime.Object, envoy.APIVersion) bool) predicate.Predicate {
+
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return filter(e.Object, version)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return filter(e.ObjectNew, version)
+
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return filter(e.Object, version)
+		},
+	}
```

Comment:
This is a filter to avoid receiving events from EnvoyConfigRevisions of the wrong envoy API version.